### PR TITLE
HealDA

### PR DESCRIPTION
**Pull request by NickGeneva**
# Earth2Studio Pull Request

## Description

- Adds HealDA data assimilation model - a new `earth2studio.models.da.HealDA` class
- Adds an example for HealDA with conventions and satellite observations
- Minor QoL improvements to UFS
- Removing surface variables from UFS due to in correct elev eval
- Addes unit tests for HealDA

```
th2studio/models/da/healda.py                335     16    95%   290, 370-371, 499, 548, 605, 643, 708, 714, 842, 848, 966-969, 977
```

Note: I need to update UFS to allow correct time tolerance ranges, same with ISD. For another PR.

Simple example:

```python
import numpy as np
from earth2studio.data import NCAR_ERA5, UFSObsConv, UFSObsSat, fetch_dataframe
from earth2studio.models.da import HealDA

package = HealDA.load_default_package()
model = HealDA.load_model(package, lat_lon=True)
model = model.to("cuda:0")

analysis_time = np.array([np.datetime64("2024-01-01T00:00")])
conv_source = UFSObsConv(tolerance=np.timedelta64(3, 'h'))
conv_schema, sat_schema = model.input_coords()
conv_df = fetch_dataframe(
    conv_source,
    time=analysis_time,
    variable=conv_schema["variable"],
    fields=np.array(list(conv_schema.keys())),
)

sat_source = UFSObsSat(tolerance=np.timedelta64(3, 'h'))
sat_df = fetch_dataframe(
    sat_source,
    time=analysis_time,
    variable=sat_schema["variable"],
    fields=np.array(list(sat_schema.keys())),
)

output = model(conv_obs=conv_df, sat_obs=sat_df)
print(output)
```

<details>

<summary>
Example render
</summary>

_home_ngeneva_Downloads__build__build69_html_examples_22_healda html (1)

</details>

<details>

<summary>
Rendered Outputs
</summary>

22_healda_obs_locations

22_healda_analysis

22_healda_differences

</details>

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/earth2studio/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/earth2studio/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/earth2studio/issues) is linked to this pull request.
- [x] Assess and address Greptile feedback (AI code review bot for guidance; use discretion, addressing all feedback is not required).

### CODE REVIEW

**Comment by NickGeneva** [comment]
@greptile-ai

File: examples/22_healda.py [hunk 1]
```diff
@@ -0,0 +1,338 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024-2026 NVIDIA CORPORATION & AFFILIATES.
```
Comment (NickGeneva) [inline examples/22_healda.py:1]: Most of the example is plotting code, see rendered result in description for the output

File: earth2studio/models/da/healda.py [hunk 2]
```diff
@@ -0,0 +1,1048 @@
...
+        -------
+        AssimilationModel
+            Loaded HealDA assimilation model
+        """
+        try:
+            package.resolve("config.json")
+        except FileNotFoundError:
+            pass
+
+        model = _HealDAModel.from_checkpoint(package.resolve("healda_ufs_era5.mdlus"))
```
[the remark this answers is not in the document]
Comment (NickGeneva) [reply earth2studio/models/da/healda.py]: This pattern is okay, config.json with hugging face packages are typically just for tracking download purposes.

File: earth2studio/data/ufs.py [hunk 3]
```diff
@@ -682,6 +687,14 @@ def _create_tasks(
...
                 logger.error(f"Variable id {v} not found in GSI satellite lexicon")
                 raise e
+            except KeyError:
+                if v in GSIConventionalLexicon:
+                    logger.warning(
+                        f"Variable id {v} is a UFS conventional variable, skipping in satellite fetch"
+                    )
+                    continue
+                logger.error(f"Variable id {v} not found in GSI lexicon")
+                raise
```
[the remark this answers is not in the document]
Comment (NickGeneva) [reply earth2studio/data/ufs.py]: Fixed.

File: earth2studio/models/da/healda.py [hunk 4]
```diff
@@ -0,0 +1,1048 @@
...
+            Time tolerance for filtering observations, by default np.timedelta64(12, "h")
+        condition : torch.Tensor | None, optional
+            Static conditioning fields (orography, land fraction) on the HEALPix grid
+            with shape [1, in_channels, time_length, npix]. If None, defaults to zeros.
+
+        Returns
+        -------
+        AssimilationModel
+            Loaded HealDA assimilation model
+        """
```
[the remark this answers is not in the document]
Comment (NickGeneva) [reply earth2studio/models/da/healda.py]: Fixed

File: earth2studio/models/da/healda.py [hunk 5]
```diff
@@ -0,0 +1,1048 @@
...
+                .values.astype("datetime64[ns]")
+                .view(np.int64),
+                "observation": df["observation"].values.astype(np.float64),
+                "local_channel": (stats["raw_to_local"][raw_ch] - 1).astype(np.int32),
+                "local_platform": df["satellite"]
+                .map(platform_map)
+                .values.astype(np.int64),
+                "sensor": sensor,
+                "obs_type": np.int32(0),
+                "height": np.float32(np.nan),
```
[the remark this answers is not in the document]
Comment (NickGeneva) [reply earth2studio/models/da/healda.py:617]: Fixed

**Comment by NickGeneva** [comment]
@greptile-ai

File: earth2studio/models/da/healda.py [hunk 6]
```diff
@@ -0,0 +1,1084 @@
...
+        batch_size = inputs["second_of_day"].shape[0]
+        noise_labels = torch.zeros([batch_size], device=self.device)
+        class_labels = torch.empty([batch_size, 0], device=self.device)
+
+        print(inputs["condition"].shape)
+        print(noise_labels.shape)
+        print(inputs["obs"].shape)
+        print(inputs["offsets"].shape)
+        print(inputs["second_of_day"].shape)
+        print(inputs["day_of_year"].shape)
```
[the remark this answers is not in the document]
Comment (NickGeneva) [reply earth2studio/models/da/healda.py]: Fixed.

File: earth2studio/models/da/healda.py [hunk 7]
```diff
@@ -0,0 +1,1084 @@
...
+            t_dt = pd.Timestamp(t).to_pydatetime()
+            midnight = t_dt.replace(hour=0, minute=0, second=0, microsecond=0)
+            jan1 = t_dt.replace(month=1, day=1, hour=0, minute=0, second=0)
+            seconds_of_day.append((t_dt - midnight).total_seconds())
+            days_of_year.append((t_dt - jan1).total_seconds() / 86400.0)
+
+        # Expand condition to match batch with the number of time inputs
+        condition = self.condition  # [batch, in_channels, time_length, npix]
+        if condition.shape[2] != n_times:
+            condition = condition[:, :, :, :].expand(n_times, -1, -1, -1).contiguous()
```
[the remark this answers is not in the document]
Comment (NickGeneva) [reply earth2studio/models/da/healda.py:897]: This is correct, n_times corresponds the batch dim here.

**Comment by NickGeneva** [comment]
@greptile-ai

File: earth2studio/models/da/healda.py [hunk 8]
```diff
@@ -0,0 +1,1077 @@
...
+                "observation": df["observation"].values.astype(np.float64),
+                "local_channel": df["variable"]
+                .map(CONV_VAR_CHANNEL)
+                .values.astype(np.int32),
+                "local_platform": np.int64(0),
+                "sensor": "conv",
+                "obs_type": df["type"].fillna(0).values.astype(np.int32),
+                "height": df["elev"].values.astype(np.float32),
+                "pressure": df["pres"].values.astype(np.float32),
+                "scan_angle": np.float32(np.nan),
```
[the remark this answers is not in the document]
Comment (NickGeneva) [reply earth2studio/models/da/healda.py:657]: Fixed.

File: earth2studio/models/da/healda.py [hunk 9]
```diff
@@ -0,0 +1,1077 @@
...
+
+        Observations are laid out **sensor-major, time-minor** in the flat
+        tensor so that all observations for a given sensor are contiguous:
+
+        ``[s0_t0, s0_t1, …, s0_tN, s1_t0, s1_t1, …, sK_tN]``
+
+        The ``offsets`` tensor (shape ``[sensors, 1, n_times]``) stores the
+        cumulative end-index in that layout, guaranteeing
+        ``offsets[s,:,:] < offsets[s+1,:,:]`` and
+        ``offsets[s,0,t] <= offsets[s,0,t+1]``.
```
[the remark this answers is not in the document]
Comment (NickGeneva) [reply earth2studio/models/da/healda.py]: Fixed.

**Comment by NickGeneva** [comment]
@greptile-ai

File: earth2studio/models/da/utils.py [hunk 10]
```diff
@@ -107,13 +110,18 @@ def filter_time_range(
...
     lower_bound, upper_bound = tolerance
-    time_min = request_time_ns + lower_bound
-    time_max = request_time_ns + upper_bound
-
-    time_mask = (df[time_column] >= time_min) & (df[time_column] <= time_max)
+    masks = [
+        (df[time_column] >= t + lower_bound) & (df[time_column] <= t + upper_bound)
+        for t in times_ns
+    ]
+    time_mask = reduce(lambda a, b: a | b, masks)
```
[the remark this answers is not in the document]
Comment (NickGeneva) [reply earth2studio/models/da/utils.py:124]: request_time will never be empty

File: earth2studio/models/da/healda.py [hunk 11]
```diff
@@ -0,0 +1,1083 @@
...
+        if unknown_vars:
+            raise ValueError(f"Unknown satellite platform(s) present: {unknown_vars}")
+
+        return pd.DataFrame(
+            {
+                "lat": df["lat"].values.astype(np.float32),
+                "lon": df["lon"].values.astype(np.float32),
+                "obs_time_ns": df["time"].values.astype("datetime64[ns]"),
+                "observation": df["observation"].values.astype(np.float64),
+                "local_channel": (stats["raw_to_local"][raw_ch] - 1).astype(np.int32),
```
[the remark this answers is not in the document]
Comment (NickGeneva) [reply earth2studio/models/da/healda.py:611]: Added

**Comment by NickGeneva** [comment]
@greptile-ai

File: test/serve/server/test_workflow.py [hunk 12]
```diff
@@ -23,21 +23,24 @@
...
+        WorkflowRegistry,
+        WorkflowResult,
+        WorkflowStatus,
+        parse_workflow_directories_from_env,
+        register_all_workflows,
+        workflow_registry,
+    )
+    from pydantic import Field, ValidationError  # type: ignore[import-untyped]
+except ImportError:
+    pass
```
[the remark this answers is not in the document]
Comment (NickGeneva) [reply test/serve/server/test_workflow.py:46]: Fixed

File: earth2studio/models/da/healda.py [hunk 13]
```diff
@@ -0,0 +1,1091 @@
...
+            lat_lon=lat_lon,
+            output_resolution=output_resolution,
+            time_tolerance=time_tolerance,
+        )
+
+    @torch.inference_mode()
+    def _forward(self, inputs: dict[str, torch.Tensor]) -> torch.Tensor:
+        batch_size = inputs["second_of_day"].shape[0]
+        noise_labels = torch.zeros([batch_size], device=self.device)
+        class_labels = torch.empty([batch_size, 0], device=self.device)
```
[the remark this answers is not in the document]
Comment (NickGeneva) [reply earth2studio/models/da/healda.py:417]: Fixed

**Comment by NickGeneva** [comment]
@greptile-ai

File: examples/22_healda.py [hunk 14]
```diff
@@ -0,0 +1,335 @@
...
+fig.subplots_adjust(wspace=0.02, hspace=0.08, left=0.1, right=0.9)
+
+lat = results[0].coords["lat"].values
+lon = results[0].coords["lon"].values
+cmaps = ["Spectral_r", "PiYG"]
+
+for row, (title, da) in enumerate(zip(titles, results)):
+    for col, var in enumerate(plot_vars):
+        ax = axes[row, col]
+        field = da.sel(variable=var).data[0].get()  # [nlat, nlon] cupy -> numpy
```
[the remark this answers is not in the document]
Comment (NickGeneva) [reply examples/22_healda.py:230]: Can assume to be cupy for this example

File: earth2studio/models/da/healda.py [hunk 15]
```diff
@@ -0,0 +1,1091 @@
...
+            package.resolve("config.json")
+        except FileNotFoundError:
+            pass
+
+        model = _HealDAModel.from_checkpoint(package.resolve("healda_ufs_era5.mdlus"))
+        model.eval()
+
+        # Static conditioning (orography, land fraction) on HEALPix grid
+        # Shape: [1, in_channels, time_length, npix]
+        if condition is None:
```
Comment (aayushg55) [inline earth2studio/models/da/healda.py]: Could we put the condition tensor into the hugging face package like the other normalization stats? There will likely be some difference between any user inputted orography/land sea mask and those used in the training depending on how regridding is done and the source of these static fields, and I can't say how sensitive the model is to changes to these inputs. The default zeros likely explains the large +-10 K t2m error (should be ~0.7). z500 error should also be ~50.

File: examples/22_healda.py [hunk 16]
```diff
@@ -0,0 +1,335 @@
...
+# The UFS data sources return pandas DataFrames that match the schemas expected by
+# :py:meth:`HealDA.input_coords`.  We use
+# :py:func:`earth2studio.data.fetch_dataframe` which attaches ``request_time``
+# metadata required by the model.  The tolerance parameter defines a 3 hour
+# window around the analysis time so that observations within that range are included.
+
+# %%
+analysis_time = np.array([np.datetime64("2024-01-01T00:00")])
+
+conv_source = UFSObsConv(tolerance=timedelta(hours=3))
```
Comment (aayushg55) [inline examples/22_healda.py]: We will update the UFS datasources to support the -21,+3 time window right?

File: examples/22_healda.py [hunk 17]
```diff
@@ -0,0 +1,335 @@
...
+                fontsize=12,
+                va="bottom",
+                ha="center",
+                rotation="vertical",
+                rotation_mode="anchor",
+                transform=ax.transAxes,
+            )
+
+fig.suptitle(
+    f"HealDA Analysis Error {str(analysis_time[0])[:16]} UTC",
```
Comment (aayushg55) [inline examples/22_healda.py:335]: We should ideally see t2m error of ~0.7k and z500 error of ~50m^2/s^2.

File: examples/22_healda.py [hunk 18]
```diff
@@ -0,0 +1,335 @@
...
+# ``create_generator`` for stateful (iterative) assimilation workflows.
+# Here we use the direct call to run three independent analyses.
+
+# %%
+torch.manual_seed(42)
+result_conv = model(conv_obs=conv_df)
+logger.info(f"Conv-only analysis shape: {result_conv.shape}")
+
+# %%
+# Run With Satellite Observations Only
```
Comment (aayushg55) [inline examples/22_healda.py]: Would be good to add a comment in the example that we'd expect very high error from just using a single source, and ideally one should use all datasources that were used in training. Smaller dropouts like excluding some combination of platforms/sensors for the satellite data would be safer if the user wants an actually useful analysis.

File: examples/22_healda.py (hunk 16 above)
Comment (NickGeneva) [reply examples/22_healda.py]: Updated

File: examples/22_healda.py (hunk 18 above)
Comment (NickGeneva) [reply examples/22_healda.py]: Added

File: earth2studio/models/da/healda.py (hunk 15 above)
Comment (NickGeneva) [reply earth2studio/models/da/healda.py]: Fixed.

File: examples/22_healda.py (hunk 18 above)
Comment (NickGeneva) [reply examples/22_healda.py]: I also change the order of the plots so the sat+conv is first since its the proper method to run the model.

File: examples/22_healda.py (hunk 17 above)
Comment (NickGeneva) [reply examples/22_healda.py:335]: 2026-03-14 00:14:09.608 | INFO     | __main__:<module>:285 - Conv+Sat - ERA5 | t2m MAE: 0.7994
2026-03-14 00:14:09.609 | INFO     | __main__:<module>:285 - Conv+Sat - ERA5 | z500 MAE: 51.2045
2026-03-14 00:14:09.610 | INFO     | __main__:<module>:285 - Sat - ERA5 | t2m MAE: 0.8488
2026-03-14 00:14:09.610 | INFO     | __main__:<module>:285 - Sat - ERA5 | z500 MAE: 164.9523
2026-03-14 00:14:09.611 | INFO     | __main__:<module>:285 - Conv - ERA5 | t2m MAE: 4.3312
2026-03-14 00:14:09.612 | INFO     | __main__:<module>:285 - Conv - ERA5 | z500 MAE: 571.0181

After update with the static fields

Comment (aayushg55) [reply examples/22_healda.py:335]: Error looks correct, verified against our HPX regridded ERA5 as well.
